Add `bignum` feature

a-rodin opened the pull request:
Ref #29 #30

So I tried to convert `BigInt` to `num_bigint::BigInt`, but it doesn't seem to be possible as the API is private. For now I've implemented `BigInt` as `i64`, using public QuickJS functions `JS_NewBigInt64` and `JS_ToBigInt64`.

I think there are three ways to eventually add support for `num_bigint::BigInt`:
1. Copy definitions of `JSBigFloat` and dependent structures to `libquickjs-sys/wrapper.h`.
2. Patch `quickjs.c` to expose some public functions for manipulations with big floats.
3. Ask QuickJS developers to expose this API in future releases.

Nonetheless, for use cases that require only `i64` this implementation would be more efficient than the one involving `num_bigint::BigInt`. So maybe even if implementation with `num_bigint::BigInt` was added, the one with `i64` could still coexist with it.

theduke commented:
Hey, thanks!
I have an idea how we could work around the API limitations and how to implement values that are `<= i64::max` more efficiently:

Add a opaque type that can hold either a `i64` or a `num_bigint::BigInt` and offer conversion methods. (See code below).

To work around the API limitations we can go through strings instead.
Rust -> JS for larger than i64::max => convert to string and call constructor (see the serialization Date handling for how to call a constructor)
JS -> Rust: convert to string and then construct the `num_biging::BigInt` with a string.

```rust
// Note: this is private to allow changing the implementation in the future
enum BigIntOrI64 {
  Int(i64),
  BigInt(num_bigint::BigInt),
}

pub struct BigInt{
  inner: BigIntOrI64,
}

impl BigInt {
  fn as_i64(&self) -> Option<i64> { ... }
  fn into_bigint(self) -> num_bigint::BigInt;
}

enum Value {
 ...
  BigInt(BigInt),
}
```

theduke commented:
Also:

- [ ] needs tests for de/serialization ( currently all in `lib.rs` )
- [ ] needs From/TryFrom impls for i64/u64/i128/u128
- [ ] We will need to store a second `embed/bindings.rs` file that includes the bindings for the BigInt functions

a-rodin commented:
@theduke I've added optional support for `num_bigint::BigInt` under `num` feature.

When it is not enabled, `JsValue::BigInt` can either be `i64` or hold a special `Overflow` values if the value returned from JS doesn't fit into `i64`.

I've implemented `From<i64>` and `From<num_bigint::BigInt>`, but I'm not sure that other implementations really need to be added, because it is not clear to be why `From<u32>` constructs `JsValue::Float`, but `From<u64>` would construct `JsValue::BigInt`. I'm not event sure that `From<i8>`, `From<u8>`, etc really need to be implemented, as there is uncertainty about which type is more appropriate to hold them.

theduke commented on libquickjs-sys/Cargo.toml:
```diff
@@ -22,5 +22,6 @@ copy_dir = { version = "0.1.2", optional = true }
 bundled = ["copy_dir"]
 patched = []
 default = ["bundled"]
+bignum = ["patched"]
```
This feature dependency should be in the main crate, not in the bindings.

theduke replied:
Actually, disregard and see my comment below.

theduke commented on src/bindings.rs:
```diff
@@ -56,6 +135,25 @@ fn js_date_constructor(context: *mut q::JSContext) -> q::JSValue {
 }
 
+#[cfg(all(feature = "bignum", feature = "num"))]
+fn js_bigint_function(context: *mut q::JSContext) -> q::JSValue {
```
```suggestion
fn js_bigint_constructor(context: *mut q::JSContext) -> q::JSValue {
```

a-rodin replied:
It is not an object constructor, because bigints are created by calling `BigInt(10)`, but not `new BigInt(10)`. So I thought that this name would highlight this difference.

theduke replied:
Function is a bit non-descriptive though.

Maybe `_create`?

theduke commented on src/bindings.rs:
```diff
@@ -400,6 +585,14 @@ fn make_cstring(value: impl Into<Vec<u8>>) -> Result<CString, ValueError> {
 }
 
+/// Helper to construct null JsValue
+fn js_null_value() -> q::JSValue {
```
I wanted to add this about 3 times already but was too lazy each time. :D 

Could probably use:
```suggestion
#[inline]
fn js_null_value() -> q::JSValue {
```

a-rodin replied:
But shouldn't compiler automatically inline it if the inlining is beneficial, as the function is used in the same file?

theduke commented on src/bindings.rs:
```diff
@@ -20,6 +22,83 @@ const TAG_UNDEFINED: i64 = 3;
 
+#[cfg(feature = "bignum")]
+#[derive(Clone, Debug)]
+pub enum BigIntOrI64 {
```
Let's move those types and impls into `value/bigint.rs`. 

(`#[cfg]` attr goes on the `mod bigint` declaration)

a-rodin replied:
I didn't put it there initially because to do it the `inner` field should be made public or more wrapping code added to allow inspection in tests.

theduke commented:
I understand the motiviation for the `Overflow` logic, but I'd really like to keep this crate easy and obvious to use. I think the overflowing could prove too confusing.

The the upstream bigint feature come with any additional platform limitations or increase compile time substantially? ( I don't think so for either)

If not, I'd suggest:
* using `quicks.bn.a` by default without a feature
* have the `BigInt` type work with `i64` and `i128` by default (with `BigIntInner::I64` and `BigIntInner::i128` variants)
* on overflow, produce a `ValueError` if the value doesn't fit into `i128`, which mentions enabling the `num_bigint_ feature
* have a `num_bigint` feature that extends number support to arbitrary length via `num_bigint::BigInt`

a-rodin commented:
If `patched` feature is not enabled, it is impossible to check for overflow, as by default QuickJS conversion to `i64` doesn't return an error, but does the conversion modulo 2^64.

a-rodin commented:
I made measurements of the test binary size compiled in release mode for different combinations of features: default, only `bignum`, and `bignum` + `num-bigint`.

The results are the following:

| Features | Size, bytes |
|---|---|
| default | 8507072 |
| `bignum` | 9295264 |
| `bignum` + `num-bigint` | 9296136 |

For this it can be seen that `num-bigint` almost doesn't increase the binary size, while just using `quickjs.bn` and adding code for handling `BigInt` does. So maybe it doesn't make sense to have two separate features, with `num-bignum` and without `num-bigint`, but have only one  feature that can be enabled by default.

theduke commented:
Agreed, I'm fine with just having a `bignum` feature.

It really doesn't make a whole lot of sense to provide anything without arbitrary support.

theduke commented:
@a-rodin what do we need to push this over the line?

Anything I can help out with?

a-rodin commented:
@theduke The problem was possibility of a memory leak, I've fixed it by using `DroppableValue` wrapper.

a-rodin commented:
@theduke Do you think it can be merged as is now?

theduke commented on libquickjs-sys/build.rs:
```diff
@@ -23,12 +28,12 @@ fn main() {
             "/usr/local/lib/quickjs"
         } else {
-            panic!("quicks is not supported on this platform");
+            panic!("quickjs is not supported on this platform");
```
Good catch. ;)

theduke commented on src/bigint.rs:
```diff
@@ -0,0 +1,65 @@
+#[cfg(feature = "num-bigint")]
```
Make a submodule under `value/bigint.rs`.

Also: only needs one `#[cfg(..)]` on the `mod bigint` declaration.

a-rodin replied:
To clarify: do you mean that the directory structure should be changed to this:

```
src/value.rs -> src/value/mod.rs
src/bigint.rs -> src/value/bigint.rs
```
?

theduke replied:
:+1:

a-rodin replied:
It is moved now

theduke commented on src/bigint.rs:
```diff
@@ -0,0 +1,65 @@
+        }
+    }
+}
+
```
```suggestion
impl Eq for BigintOrI64 {}
```

theduke commented on src/bindings.rs:
```diff
@@ -206,6 +229,54 @@ fn serialize_value(context: *mut q::JSContext, value: JsValue) -> Result<q::JSVa
+        #[cfg(feature = "num-bigint")]
+        JsValue::BigInt(int) => match int.inner {
+            BigIntOrI64::Int(int) => unsafe { q::JS_NewBigInt64(context, int) },
+            #[cfg(feature = "num-bigint")]
```
This `cfg` seems redundant.

theduke commented on src/bindings.rs:
```diff
@@ -413,6 +484,41 @@ fn deserialize_value(
+                if ret == -1 {
+                    return Err(ValueError::Internal("Internal BigInt error".into()));
+                }
+                #[cfg(feature = "num-bigint")]
```
also redundant

theduke commented on src/bindings.rs:
```diff
@@ -413,6 +484,41 @@ fn deserialize_value(
+                }))
+            } else {
+                if ret == -1 {
+                    return Err(ValueError::Internal("Internal BigInt error".into()));
```
```suggestion
                    return Err(ValueError::Internal("BigInt conversion error".into()));
```

Edit: actually, should this if still be there?

a-rodin replied:
Yeah, I think this check is not needed at this point

theduke commented on src/bindings.rs:
```diff
@@ -999,3 +1108,44 @@ impl ContextWrapper {
 }
+
+#[cfg(test)]
+mod tests {
```
Move to in `value/bigint.rs`.

theduke commented on src/value.rs:
```diff
@@ -93,6 +97,20 @@ value_impl_from! {
 }
 
+#[cfg(feature = "num-bigint")]
+impl From<i64> for JsValue {
```
Probably will also want a `From<u64/i128/u128>`

a-rodin replied:
It is added now

theduke commented:
Also missing: `TryFrom` impls for `Value -> Rust type` conversion. Otherwise `eval_as` won't work. 
Also: a test exercising `eval_as`.

a-rodin commented on Cargo.toml:
```diff
@@ -13,6 +13,8 @@ keywords = ["quickjs", "javascript", "js", "engine", "interpreter"]
 [dependencies]
 libquickjs-sys = { version = "0.3.0", path = "./libquickjs-sys" }
 chrono = { version = "0.4.7", optional = true }
+num-bigint-bare = { package = "num-bigint", version = "0.2.2", optional = true }
```
It was renamed to `-bare` to enable feature named `num-bigint`, which is not needed now

theduke commented:
Awesome, thanks for working on this!

theduke commented:
Rebased and pushed to master!
